clients/src/main/java/org/apache/kafka/clients/NetworkClient.java

```diff
@@ -440,6 +440,7 @@ private void handleCompletedReceives(List<ClientResponse> responses, long now) {
             correlate(req.request().header(), header);
             if (!metadataUpdater.maybeHandleCompletedReceive(req, now, body))
                 responses.add(new ClientResponse(req, now, false, body));
+            inFlightRequests.completeNext(source);
```

Comment:
@junrao I replied on the ticket. I think the broker side bug has been fixed after KAFKA-2071. So the broker will always send an error ProduceResponse in the same version of the ProduceRequest. Previously it seems always sending the latest version in ```handleError()```.

When this issue happened, the client logged an exception in the sender thread and moved on. I agree it is not a problem on producer side if the broker can always do things right. I'm wondering if we would want the producer itself to be resilient enough to handle this case?